database/tbcd/level/level.go

```diff
@@ -411,70 +436,75 @@ func (l *ldb) BlockHeadersInsert(ctx context.Context, bhs [][80]byte) (tbcd.Inse
 	// headers on what the network may be doing. Not sure how to handle
 	// that right now but leaving a note.
 
-	if len(bhs) == 0 {
-		return tbcd.ITInvalid, nil, nil,
-			errors.New("block headers insert: no block headers to insert")
+	if len(bhs.Headers) == 0 {
+		return tbcd.ITInvalid, nil, nil, 0,
+			errors.New("block headers insert: invalid")
 	}
 
 	// Ensure we can connect these blockheaders prior to starting database
 	// transaction. This also obtains the starting cumulative difficulty
 	// and  height.
-	wbh, err := tbcd.B2H(bhs[0][:])
-	if err != nil {
-		return tbcd.ITInvalid, nil, nil,
-			fmt.Errorf("block headers insert b2h: %w", err)
-	}
+	wbh := bhs.Headers[0]
 	pbh, err := l.BlockHeaderByHash(ctx, &wbh.PrevBlock)
 	if err != nil {
-		return tbcd.ITInvalid, nil, nil,
+		return tbcd.ITInvalid, nil, nil, 0,
 			fmt.Errorf("block headers insert: %w", err)
 	}
 
 	// block headers
 	bhsTx, bhsCommit, bhsDiscard, err := l.startTransaction(level.BlockHeadersDB)
 	if err != nil {
-		return tbcd.ITInvalid, nil, nil,
+		return tbcd.ITInvalid, nil, nil, 0,
 			fmt.Errorf("block headers open transaction: %w", err)
 	}
 	defer bhsDiscard()
 
-	// Make sure we are not inserting the same blocks
-	bhash := wbh.BlockHash()
-	has, err := bhsTx.Has(bhash[:], nil)
-	if err != nil {
-		return tbcd.ITInvalid, nil, nil,
-			fmt.Errorf("block headers insert has: %w", err)
+	// Iterate over the block headers and skip block headers we already
+	// have in the database. Rely on caching to make this not suck terribly.
+	var x int
+	for _, rbh := range bhs.Headers {
+		bhash := rbh.BlockHash()
+		has, err := bhsTx.Has(bhash[:], nil)
+		if err != nil {
+			return tbcd.ITInvalid, nil, nil, 0,
+				fmt.Errorf("block headers insert has: %w", err)
+		}
+		if has {
+			x++
+		} else {
+			break
+		}
```

Comment:
I think this could be simplified:
```suggestion
		if !has {
			break
		}
		x++
```